### ISSUE TITLE
mitmproxy fails to start with message "cannot connect to X server"

### ISSUE THREAD

**Issue by chekunkov** [mitmproxy/mitmproxy#701]
just upgraded from v0.9.2 to v0.13

```
sudo pip install mitmproxy --upgrade
```

and now mitmproxy doesn't start, I'm getting the following message:

```
# vagrant at dev in ~ [12:34:26]
$ mitmproxy
: cannot connect to X server

# vagrant at dev in ~ [12:34:29]
$
```

mitmproxy version: 0.13
operating system: Ubuntu 14.04.1 LTS

**Comment by chekunkov** [comment]
works again only after downgrade to v0.11.2

```
sudo pip install mitmproxy==0.11.2 --upgrade
```

v0.12.1 fails with the same message and v0.12.0 raises `ImportError: cannot import name socks` so I wasn't able to test it.

I would like to use new features from v0.13 so any advice is welcome.

**Comment by mhils** [comment]
Hm, interesting. What kind of console/host os are you using? How can we reproduce this bug?

**Comment by chekunkov** [comment]
Let me provide some more details:

Host OS: Mac OS X 10.10.4
Vagrant: 4.3.20
Oracle VM VirtualBox Manager 4.3.20
Guest OS: Ubuntu 14.04.1 LTS
Python 2.7.6

Initially I've tried to install mitmproxy to the system packages, this results in:

```
# vagrant at dev in ~ [19:08:54]
$ mitmproxy
: cannot connect to X server

# vagrant at dev in ~ [19:09:04]
$ sudo mitmproxy
: cannot connect to X server
```

`sudo pip freeze` output:

```
ansible==1.5.4
apt-xapian-index==0.45
backports.ssl-match-hostname==3.4.0.2
blinker==1.4
boto==2.34.0
certifi==2015.4.28
cffi==1.1.2
chardet==2.0.1
Cheetah==2.4.4
cloud-init==0.7.5
ConfigArgParse==0.9.3
configobj==4.7.2
cryptography==0.9.3
cssselect==0.9.1
Cython==0.22.1
decorator==3.4.2
distro-info==0.12
ecdsa==0.11
enum34==1.0.4
euca2ools==3.0.1
Fabric==1.10.0
Flask==0.10.1
functools32==3.2.3.post1
gyp==0.1
h5py==2.5.0
hpack==1.1.0
html2text==2015.6.21
httpie==0.8.0
httplib2==0.8
idna==2.0
ipaddress==1.0.14
ipdb==0.8
ipython==3.2.0
itsdangerous==0.24
Jinja2==2.8
jsonpatch==1.3
jsonpointer==1.0
jsonschema==2.5.1
keyring==3.5
Landscape-Client==14.1
launchpadlib==1.10.2
lazr.restfulclient==0.13.3
lazr.uri==1.0.3
leveldb==0.193
lxml==3.4.4
MarkupSafe==0.23
matplotlib==1.4.3
meld3==1.0.0
mercurial==2.8.2
mitmproxy==0.13
mock==1.0.1
netlib==0.13.1
networkx==1.9.1
nose==1.3.7
numpy==1.9.2
oauth==1.0.1
PAM==0.4.2
pandas==0.16.2
paramiko==1.15.1
passlib==1.6.4
pbr==0.10.0
Pillow==2.9.0
prettytable==0.7.2
protobuf==2.6.1
py==1.4.26
pyasn1==0.1.8
pycparser==2.14
pycrypto==2.6.1
pycurl==7.19.3
Pygments==1.6
pygobject==3.12.0
pyinotify==0.9.4
pyOpenSSL==0.15.1
pyparsing==2.0.3
pyperclip==1.5.11
pyserial==2.6
python-apt==0.9.3.5
python-dateutil==1.5
python-debian===0.1.21-nmu2ubuntu2
python-gflags==2.0
pytz==2015.4
PyYAML==3.10
pyzmq==14.7.0
queuelib==1.2.2
requestbuilder==0.1.0b1
requests==2.2.1
scikit-image==0.11.3
scipy==0.15.1
Scrapy==1.0.0
SecretStorage==2.0.0
simplejson==3.3.1
six==1.9.0
sortedcontainers==0.9.6
ssh-import-id==3.21
stevedore==1.1.0
supervisor==3.1.3
tornado==4.2.1
tox==1.8.1
Twisted==15.2.1
Twisted-Core==13.2.0
Twisted-Names==13.2.0
Twisted-Web==13.2.0
urllib3==1.7.1
urwid==1.3.0
virtualenv==1.11.6
virtualenv-clone==0.2.5
virtualenvwrapper==4.3.1
w3lib==1.11.0
wadllib==1.3.2
Werkzeug==0.10.4
zope.interface==4.0.5
```

Today I've tried to install 0.13 to a virtualenv

```
# vagrant at dev in ~ [19:03:20]
$ mkvirtualenv mitmproxy-test
New python executable in mitmproxy-test/bin/python
Installing setuptools, pip...done.

(mitmproxy-test)
# vagrant at dev in ~ [19:03:35]
$ pip install mitmproxy==0.13
```

Which resulted in:

```
(mitmproxy-test)
# vagrant at dev in ~ [19:12:16]
$ mitmproxy
Error starting proxy server: error(98, 'Address already in use')

(mitmproxy-test)
# vagrant at dev in ~ [19:15:31]
$ sudo mitmproxy
: cannot connect to X server
```

`pip freeze` output:

```
ConfigArgParse==0.9.3
Pillow==2.9.0
argparse==1.2.1
backports.ssl-match-hostname==3.4.0.2
blinker==1.4
certifi==2015.04.28
cffi==1.1.2
cryptography==0.9.3
enum34==1.0.4
hpack==1.1.0
html2text==2015.6.21
idna==2.0
ipaddress==1.0.14
lxml==3.4.4
mitmproxy==0.13
netlib==0.13.1
passlib==1.6.4
pyOpenSSL==0.15.1
pyasn1==0.1.8
pycparser==2.14
pyparsing==2.0.3
pyperclip==1.5.11
six==1.9.0
tornado==4.2.1
urwid==1.3.0
wsgiref==0.1.2
```

As mitmproxy complained about port in use - I've checked that

```
(mitmproxy-test)
# vagrant at dev in ~ [19:20:02]
$ sudo netstat | grep 98
(mitmproxy-test)
# vagrant at dev in ~ [19:20:26]
$
```

And just in case - I have [Anaconda](http://continuum.io/downloads) installed - with the following list of packages:

```
$ conda list
# packages in environment at /home/vagrant/anaconda:
#
You are using pip version 7.0.3, however version 7.1.0 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
_license                  1.1                      py27_0
abstract-rendering        0.5.1                np19py27_0
alabaster                 0.7.3                    py27_0
anaconda                  2.3.0                np19py27_0
argcomplete               0.8.9                    py27_0
astropy                   1.0.3                np19py27_0
atlas                     3.8.4                         0
babel                     1.3                      py27_0
backports.ssl-match-hostname 3.4.0.2                   <pip>
bcolz                     0.9.0                np19py27_0
beautiful-soup            4.3.2                    py27_0
beautifulsoup4            4.3.2                     <pip>
binstar                   0.11.0                   py27_0
bitarray                  0.8.1                    py27_0
blaze                     0.8.0                     <pip>
blaze-core                0.8.0                np19py27_0
blz                       0.6.2                np19py27_1
bokeh                     0.9.0                np19py27_0
boost                     1.57.0                        4
boto                      2.38.0                   py27_0
bottleneck                1.0.0                np19py27_0
cairo                     1.12.18                       4
cdecimal                  2.3                      py27_0
certifi                   14.05.14                 py27_0
cffi                      1.1.0                    py27_0
clyent                    0.3.4                    py27_0
colorama                  0.3.3                    py27_0
conda                     3.14.1                   py27_0
conda-build               1.14.1                   py27_0
conda-env                 2.2.3                    py27_0
configobj                 5.0.6                    py27_0
cryptography              0.9.1                    py27_0
cython                    0.22.1                   py27_0
cytoolz                   0.7.3                    py27_0
datashape                 0.4.5                np19py27_0
decorator                 3.4.2                    py27_0
docutils                  0.12                     py27_0
dynd-python               0.6.5                np19py27_0
enum34                    1.0.4                    py27_0
fastcache                 1.0.2                    py27_0
flask                     0.10.1                   py27_1
fontconfig                2.11.1                        4
freetype                  2.5.2                         2
funcsigs                  0.4                      py27_0
gevent                    1.0.1                    py27_0
gevent-websocket          0.9.3                    py27_0
greenlet                  0.4.7                    py27_0
grin                      1.2.1                    py27_1
h5py                      2.5.0                np19py27_2
hdf5                      1.8.14                        0
icu                       54.1                          0
idna                      2.0                      py27_0
ipaddress                 1.0.7                    py27_0
ipython                   3.2.0                    py27_0
ipython-notebook          3.2.0                    py27_0
ipython-qtconsole         3.2.0                    py27_0
itsdangerous              0.24                     py27_0
jdcal                     1.0                      py27_0
jedi                      0.8.1                    py27_0
jinja2                    2.7.3                    py27_1
jpeg                      8d                            0
jsonschema                2.4.0                    py27_0
leveldb                   1.15                          1
libdynd                   0.6.5                         0
libffi                    3.0.13                        0
libpng                    1.6.17                        0
libsodium                 0.4.5                         0
libtiff                   4.0.2                         1
libxml2                   2.9.2                         0
libxslt                   1.1.28                        0
llvmlite                  0.5.0                    py27_0
lxml                      3.4.4                    py27_0
markupsafe                0.23                     py27_0
matplotlib                1.4.3                np19py27_2
mistune                   0.5.1                    py27_1
mkl-rt                    11.1                         p0
mkl-service               1.0.0                   py27_p1
mock                      1.0.1                    py27_0
multipledispatch          0.4.7                    py27_0
networkx                  1.9.1                    py27_0
nltk                      3.0.3                np19py27_0
nose                      1.3.7                    py27_0
numba                     0.19.1               np19py27_0
numexpr                   2.3.1                np19py27_0
numpy                     1.9.2                    py27_0
odo                       0.3.2                np19py27_0
opencv                    2.4.10               np19py27_1
openpyxl                  1.8.5                    py27_0
openssl                   1.0.1k                        1
pandas                    0.16.2               np19py27_0
patsy                     0.3.0                np19py27_0
pep8                      1.6.2                    py27_0
pillow                    2.8.2                    py27_0
pip                       7.0.3                    py27_0
pixman                    0.26.2                        0
ply                       3.6                      py27_0
protobuf                  2.6.1                    py27_1
psutil                    2.2.1                    py27_0
ptyprocess                0.4                      py27_0
py                        1.4.27                   py27_0
py2cairo                  1.10.0                   py27_2
pyasn1                    0.1.7                    py27_0
pycosat                   0.6.1                    py27_0
pycparser                 2.14                     py27_0
pycrypto                  2.6.1                    py27_0
pycurl                    7.19.5.1                 py27_2
pyflakes                  0.9.2                    py27_0
pygments                  2.0.2                    py27_0
pyopenssl                 0.15.1                   py27_1
pyparsing                 2.0.3                    py27_0
pyqt                      4.11.3                   py27_1
pytables                  3.2.0                np19py27_0
pytest                    2.7.1                    py27_0
python                    2.7.10                        0
python-dateutil           2.4.2                    py27_0
python-gflags             2.0                      py27_0
pytz                      2015.4                   py27_0
pyyaml                    3.11                     py27_1
pyzmq                     14.7.0                   py27_0
qt                        4.8.6                         3
readline                  6.2                           2
redis                     2.6.9                         0
redis-py                  2.10.3                   py27_0
requests                  2.7.0                    py27_0
rope                      0.9.4                    py27_1
runipy                    0.1.3                    py27_0
scikit-image              0.11.3               np19py27_0
scikit-learn              0.16.1               np19py27_0
scipy                     0.15.1               np19py27_0
setuptools                17.1.1                   py27_0
sip                       4.16.5                   py27_0
six                       1.9.0                    py27_0
snappy                    1.1.1                         0
snowballstemmer           1.2.0                    py27_0
sockjs-tornado            1.0.1                    py27_0
sphinx                    1.3.1                    py27_0
sphinx-rtd-theme          0.1.7                     <pip>
sphinx_rtd_theme          0.1.7                    py27_0
spyder                    2.3.5.2                  py27_0
spyder-app                2.3.5.2                  py27_0
sqlalchemy                1.0.5                    py27_0
sqlite                    3.8.4.1                       1
ssl_match_hostname        3.4.0.2                  py27_0
statsmodels               0.6.1                np19py27_0
sympy                     0.7.6                    py27_0
system                    5.8                           2
tables                    3.2.0                     <pip>
terminado                 0.5                      py27_0
theano                    0.7.0                np19py27_0
tk                        8.5.18                        0
toolz                     0.7.2                    py27_0
tornado                   4.2                      py27_0
ujson                     1.33                     py27_0
unicodecsv                0.9.4                    py27_0
util-linux                2.21                          0
werkzeug                  0.10.4                   py27_0
xlrd                      0.9.3                    py27_0
xlsxwriter                0.7.3                    py27_0
xlwt                      1.0.0                    py27_0
yaml                      0.1.6                         0
zeromq                    4.0.5                         0
zlib                      1.2.8                         0
```

Just tell me if I can give you even more information

**Comment by Grazfather** [comment]
+1

**Comment by milesrichardson** [comment]
Also seeing this on ubuntu 14.04, the issue seems to be with pyperclip

**Comment by mhils** [comment]
@milesrichardson, thanks. How do you conclude that it's pyperclip?

**Comment by milesrichardson** [comment]
I have been installing a certain version of mitmproxy on boxes for the past couple months, and it has always worked -- it only started failing very recently (possibly due to some weird configuration on my box)

```
$ sudo pip install git+git://github.com/mitmproxy/mitmproxy.git@946fd375fa87fb8a4d071a354fb99840b0ccfdf3 --upgrade
Collecting git+git://github.com/mitmproxy/mitmproxy.git@946fd375fa87fb8a4d071a354fb99840b0ccfdf3
  Cloning git://github.com/mitmproxy/mitmproxy.git (to 946fd375fa87fb8a4d071a354fb99840b0ccfdf3) to /tmp/pip-MNQC2H-build
  Could not find a tag or branch '946fd375fa87fb8a4d071a354fb99840b0ccfdf3', assuming commit.
Collecting pyperclip>=1.5.8 (from mitmproxy==0.12.1)
  Using cached pyperclip-1.5.17.zip
    Complete output from command python setup.py egg_info:
    : cannot connect to X server
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-hgTASs/pyperclip
```

It's somehow related to `sudo` and not having a `DISPLAY` variable set. I don't know why X11 forwarding or an X server is required at all...

Interestingly when I downgrade mitmproxy to `0.11.2`, the screen is not black when I attach with `mitmproxy`, but rather appears to be native terminal, and many interactions are broken (cannot use keyboard shortcuts between panes for example). 

I am on ubuntu 14.04 ec2 in a vpc with source/dest checking enabled and connected via l2tp vpn.

**Comment by mhils** [comment]
Thanks for the feedback!
`pyperclip` is an optional dependency, you can safely skip it.

**Comment by gato** [comment]
Also in linux pyperclip has 5 possible methods of accessing the clipboard: xclip, xsel,gtk , PyQt4 module or klipper (new one)

In pyperclip 1.5.11 the selection order was xclip, gtk , pyqt, xsel but in master the order is gtk, pyqt, xclip, xsel, klipper. 

Maybe using a different clipboard doesn't raises the issue and still allows copy to clipboard functionality

**Comment by milesrichardson** [comment]
I installed all of those libraries and none of them helped... I think the issue is upstream (i.e., related to the non-presence of an X-Server). I imagine this will be a problem with most EC2 servers?

**Comment by Grazfather** [comment]
ec2 is where I had the troubles. It's definitely a regression and that should be enough reason to fix.

**Comment by mhils** [comment]
While we're tracking this down in pyperclip, the workaround for this issue is to update pyperclip to 1.5.19 (`pip install --upgrade pyperclip`) and run `apt-get install python-gtk2`.

**Comment by Grazfather** [comment]
the pip install actually failed... installing `python-gtk2` first _then_ pyperclip worked for me. Thanks!

**Comment by mhils** [comment]
Thanks to the awesome @asweigart, this should be fixed now! Thank you everyone for reporting. :smiley:

**Comment by Grazfather** [comment]
Fixed with asweigart/pyperclip@56f161410d6 (asweigart/pyperclip#36)
